File: leetcode/Sort/leetcode-164/164.java
class Solution {
    public int maximumGap(int[] nums) {
        if (nums == null || nums.length < 2) {
            return 0;
        }

        int max = nums[0];
        int min = nums[0];
        int len = nums.length;
        for (int num : nums) {
            max = Math.max(max, num);
            min = Math.min(min, num);
        }

        int bucketCap = (max - min) / len + 1;
        int bucketCnt = (max - min) / bucketCap + 1;
        int[] bucketMax = new int[bucketCnt];
        int[] bucketMin = new int[bucketCnt];
        Arrays.fill(bucketMax, Integer.MIN_VALUE);
        Arrays.fill(bucketMin, Integer.MAX_VALUE);

        HashSet<Integer> set = new HashSet<>();

        for (int num : nums) {
            int idx = (num - min) / bucketCap;
            bucketMax[idx] = Math.max(bucketMax[idx], num);
            bucketMin[idx] = Math.min(bucketMin[idx], num);
            set.add(idx);
        }

        int pre = 0, res = 0;
        while (!set.contains(pre)) {
            pre++;
        }

        for (int i = pre + 1; i < bucketCnt; i++) {
            if (!set.contains(i)) {
                continue;
            }
            res = Math.max(res, bucketMin[i] - bucketMax[pre]);
            pre = i;
        }

        return res;
    }
}

File: leetcode/Sort/leetcode-164/164.go
package sort

import "math"

func maximumGap(nums []int) int {
	if nums == nil || len(nums) < 2 {
		return 0
	}

	l, max, min := len(nums), nums[0], nums[0]

	for _, num := range nums {
		if num > max {
			max = num
		}
		if num < min {
			min = num
		}
	}

	bCap := (max-min)/l + 1
	bCnt := (max-min)/bCap + 1
	bMax := make([]int, bCnt)
	bMin := make([]int, bCnt)
	for i := 0; i < bCnt; i++ {
		bMax[i] = math.MinInt32
		bMin[i] = math.MaxInt32
	}

	set := map[int]bool{}

	for _, num := range nums {
		idx := (num - min) / bCap
		if num > bMax[idx] {
			bMax[idx] = num
		}
		if num < bMin[idx] {
			bMin[idx] = num
		}
		set[idx] = true
	}

	pre, res := 0, 0
	for _, ok := set[pre]; !ok; pre++ {
	}

	for i := pre + 1; i < bCnt; i++ {
		if _, ok := set[i]; !ok {
			continue
		}
		if res < bMin[i]-bMax[pre] {
			res = bMin[i] - bMax[pre]
		}
		pre = i
	}
	return res
}
